rclrs/src/parameter/service.rs

```diff
@@ -1,89 +1,941 @@
-use std::sync::{Arc, Weak, Mutex};
+use std::collections::BTreeSet;
+use std::sync::{Arc, Mutex};
 
-use crate::vendor::rcl_interfaces::srv::rmw::*;
 use crate::vendor::rcl_interfaces::msg::rmw::*;
-use rosidl_runtime_rs::{Sequence, seq};
+use crate::vendor::rcl_interfaces::srv::rmw::*;
+use rosidl_runtime_rs::Sequence;
 
-use crate::{rmw_request_id_t, Node, RclrsError, Service, ServiceBase};
-use crate::rcl_bindings::rcl_node_t;
+use super::ParameterMap;
+use crate::parameter::{DeclaredValue, ParameterKind, ParameterStorage};
+use crate::{rmw_request_id_t, Node, RclrsError, Service};
 
+// The variables only exist to keep a strong reference to the services and are technically unused.
+// What is used is the Weak that is stored in the node, and is upgraded when spinning.
 pub struct ParameterService {
+    #[allow(dead_code)]
     describe_parameters_service: Arc<Service<DescribeParameters>>,
```

Comment:
I'm not really understanding this, shouldn't the executor eventually need handles to these services?

Reply:
I'm a bit fuzzy on the details since it's been a few months since I wrote this but I believe that the following happens:

* You call [create_service](https://github.com/ros2-rust/ros2_rust/blob/875bda3855d748e6169457592af198211f5d2af5/rclrs/src/node.rs#L274-L291), it returns a strong reference but holds a weak reference internally.
* The executor's spin_once call creates a new [WaitSet](https://github.com/ros2-rust/ros2_rust/blob/875bda3855d748e6169457592af198211f5d2af5/rclrs/src/executor.rs#L50) that waits and collects ready entities.
* The `new_for_node` wait set function, collects the subscriptions by using the [live_*](https://github.com/ros2-rust/ros2_rust/blob/875bda3855d748e6169457592af198211f5d2af5/rclrs/src/node.rs#L319-L345) functions, that iterate through all Weak objects and try to upgrade them to a strong reference, returning all valid pointers.

The result I believe, is that the caller of `create_service` _must_ hold on to a strong reference otherwise that service will be dropped. However, the node itself will store its own `Weak` reference and the strong reference will actually never be used at all, but we must still be sure to keep the service alive for the necessary time.

Reply:
Ahh, I see! Thanks for the clarification. 

Is this TODO still accurate then? Seems like because we will drop these strong references when the node is dropped that our lifetimes are aligned.
https://github.com/ros2-rust/ros2_rust/blob/875bda3855d748e6169457592af198211f5d2af5/rclrs/src/node.rs#L273

Reply:
I haven't introduced that TODO but for my understanding it would still be accurate.
Since the strong reference is returned by the `create_*` call but the node only holds a weak one, what could actually happen is something like this (pathological case):

```
let mut sub;
{
    let node = create_node([...]);
    sub = node.create_subscription([...]);
    // Node is dropped!
}
// sub still has an owning reference, what now?
```